pkg/ruler/mapper_test.go

```diff
@@ -382,16 +466,145 @@ func Test_mapper_MapRulesSpecialCharNamespace(t *testing.T) {
 	})
 }
 
-func Test_mapper_CleanupShouldNotFailIfPathDoesNotExist(t *testing.T) {
+func Test_mapper_users(t *testing.T) {
+	l := util_log.MakeLeveledLogger(os.Stdout, "info")
+	setupRuleSets()
 	m := &mapper{
-		Path:   "/path-does-not-exist",
+		Path:   "/rules",
 		FS:     afero.NewMemMapFs(),
-		logger: log.NewNopLogger(),
+		logger: l,
+	}
+
+	t.Run("should not fail if path does not exist", func(t *testing.T) {
+		m := &mapper{
+			Path:   "/path-does-not-exist",
+			FS:     afero.NewMemMapFs(),
+			logger: log.NewNopLogger(),
+		}
+
+		actual, err := m.users()
+		require.NoError(t, err)
+		require.Empty(t, actual)
+	})
+
+	t.Run("adding a rulegroup returns the user", func(t *testing.T) {
+		_, _, err := m.MapRules(testUser, initialRuleSet)
+		require.NoError(t, err)
+
+		result, err := m.users()
+
+		require.NoError(t, err)
+		require.Len(t, result, 1)
+		require.Contains(t, result, testUser)
+	})
+
+	t.Run("adding a rulegroup for a second user returns both users", func(t *testing.T) {
+		_, _, err := m.MapRules(testUser2, initialRuleSet)
+		require.NoError(t, err)
+
+		result, err := m.users()
+
+		require.NoError(t, err)
+		require.Len(t, result, 2)
+		require.Contains(t, result, testUser)
+		require.Contains(t, result, testUser2)
+	})
+
+	t.Run("deleting a user's rule groups keeps that user", func(t *testing.T) {
+		_, _, err := m.MapRules(testUser, map[string][]rulefmt.RuleGroup{})
+		require.NoError(t, err)
+
+		// This happens because MapRules does not delete the user directory if it cleared all the files inside.
+		// However, users() only looks at the set of user directories.
+		// This is something that can be improved on in the future.
+		result, err := m.users()
+
+		require.NoError(t, err)
+		require.Len(t, result, 2)
+		require.Contains(t, result, testUser)
+		require.Contains(t, result, testUser2)
+	})
+
+	t.Run("cleanup removes all users", func(t *testing.T) {
+		m.cleanup()
+
+		result, err := m.users()
+
+		require.NoError(t, err)
+		require.Empty(t, result)
+	})
+}
+
+func Test_FSLoader_LoadRules(t *testing.T) {
+	l := util_log.MakeLeveledLogger(os.Stdout, "info")
+	setupRuleSets()
+	fs := afero.NewMemMapFs()
+	m := &mapper{
+		Path:   "/rules",
+		FS:     fs,
+		logger: l,
 	}
 
-	actual, err := m.users()
-	require.NoError(t, err)
-	require.Empty(t, actual)
+	t.Run("basic rulegroup", func(t *testing.T) {
+		updated, files, err := m.MapRules(testUser, initialRuleSet)
+		require.True(t, updated)
+		require.Len(t, files, 1)
+		require.Equal(t, fileOnePath, files[0])
+		require.NoError(t, err)
+
+		loader := NewFSLoader(fs)
+		loaded, errs := loader.Load(fileOnePath, false, model.LegacyValidation)
+		require.Empty(t, errs)
+		require.NotNil(t, loaded)
+		require.Len(t, loaded.Groups, 2)
+		// Groups are sorted in reverse order by name, so "two" comes before "one".
```

Comment:
This seems a bit weird, but that's really what the mapper does:
https://github.com/grafana/mimir/blob/1700034d3b112052e133bdb1940cd29ab521c236/pkg/ruler/mapper.go#L146-L148
I suspect the reversal of `a` and `b` is a very old typo. Ultimately it doesn't matter too much as long as it's consistent, so I kept it.